accounts/keystore/keystore_passphrase.go

```diff
@@ -106,9 +107,79 @@ func (ks keyStorePassphrase) JoinPath(filename string) string {
 	return filepath.Join(ks.keysDirPath, filename)
 }
 
+// encryptCrypto encrypts a private key to a cryptoJSON object.
+func encryptCrypto(keyBytes []byte, auth string, scryptN, scryptP int) (*cryptoJSON, error) {
+	authArray := []byte(auth)
+
+	salt := make([]byte, 32)
+	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
+		panic("reading from crypto/rand failed: " + err.Error())
+	}
+	derivedKey, err := scrypt.Key(authArray, salt, scryptN, scryptR, scryptP, scryptDKLen)
+	if err != nil {
+		return nil, err
+	}
+	encryptKey := derivedKey[:16]
+
+	iv := make([]byte, aes.BlockSize) // 16
+	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
+		panic("reading from crypto/rand failed: " + err.Error())
+	}
+	cipherText, err := aesCTRXOR(encryptKey, keyBytes, iv)
+	if err != nil {
+		return nil, err
+	}
+	mac := crypto.Keccak256(derivedKey[16:32], cipherText)
+
+	scryptParamsJSON := make(map[string]interface{}, 5)
+	scryptParamsJSON["n"] = scryptN
+	scryptParamsJSON["r"] = scryptR
+	scryptParamsJSON["p"] = scryptP
+	scryptParamsJSON["dklen"] = scryptDKLen
+	scryptParamsJSON["salt"] = hex.EncodeToString(salt)
+
+	cipherParamsJSON := cipherparamsJSON{
+		IV: hex.EncodeToString(iv),
+	}
+
+	return &cryptoJSON{
+		Cipher:       "aes-128-ctr",
+		CipherText:   hex.EncodeToString(cipherText),
+		CipherParams: cipherParamsJSON,
+		KDF:          keyHeaderKDF,
+		KDFParams:    scryptParamsJSON,
+		MAC:          hex.EncodeToString(mac),
+	}, nil
+}
+
 // EncryptKey encrypts a key using the specified scrypt parameters into a json
-// blob that can be decrypted later on.
+// blob that can be decrypted later on. It uses the keystore v4 format.
 func EncryptKey(key Key, auth string, scryptN, scryptP int) ([]byte, error) {
+	pks := key.GetPrivateKeys()
+	crypto := make([][]cryptoJSON, len(pks))
+	for i, keys := range pks {
+		crypto[i] = make([]cryptoJSON, len(keys))
+		for j, k := range keys {
+			keyBytes := math.PaddedBigBytes(k.D, 32)
+			c, err := encryptCrypto(keyBytes, auth, scryptN, scryptP)
+			if err != nil {
+				return nil, err
+			}
+			crypto[i][j] = *c
+		}
+	}
+	encryptedKeyJSONV4 := encryptedKeyJSONV4{
+		hex.EncodeToString(key.GetAddress().Bytes()),
+		crypto,
+		key.GetId().String(),
+		4,
+	}
+	return json.Marshal(encryptedKeyJSONV4)
+}
+
+// EncryptKeyV3 encrypts a key using the specified scrypt parameters into a json
+// blob that can be decrypted later on. It uses the keystore v3 format.
+func EncryptKeyV3(key Key, auth string, scryptN, scryptP int) ([]byte, error) {
```

Comment:
In this function, why don't you use `encryptCrypto`? I think it is reusable.

Reply:
Good idea. I thought about that, but I thought it makes code change larger (it is already large, though).
I will do that after this PR is merged. :)